pkg/kubelet/dockershim/checkpoint_store.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockershim
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+const (
+	tmpPrefix    = "."
+	tmpSuffix    = ".tmp"
+	keyMaxLength = 250
+	keyRegex     = "^[a-zA-Z0-9]+$"
+)
+
+// CheckpointStore provides the interface for checkpoint storage backend
+type CheckpointStore interface {
+	// key must contain one or more characters in [A-Za-z0-9]
+	// Write persists a checkpoint with key
+	Write(key string, data []byte) error
+	// Read retrieves a checkpoint with key
+	Read(key string) ([]byte, error)
+	// Delete deletes a checkpoint with key
+	Delete(key string) error
+	// List lists all keys of existing checkpoints
+	List() ([]string, error)
+}
+
+// FileStore is an implementation of CheckpointStore interface which stores checkpoint in file.
+type FileStore struct {
+	path string
+}
+
+func (fstore *FileStore) Write(key string, data []byte) error {
+	if !isValidKey(key) {
+		return fmt.Errorf("Checkpoint key %q is not valid.", key)
+	}
+	if _, err := os.Stat(fstore.path); os.IsNotExist(err) {
+		if err = os.MkdirAll(fstore.path, 0755); err != nil && !os.IsExist(err) {
+			return err
+		}
+	}
```

Comment:
> If err == nil or the directory already exists, move on; otherwise, return an error.

Why are we sure the directory exist when the error is not `IsNotExist`? Is it possible that we just hit some fs error that the function returns error even before check the existence?